// git-info-webpack-plugin/index.js
const { handleCompiler, runGit } = require('./utils');

const COMMITHASH_COMMAND = 'rev-parse HEAD';
const VERSION_COMMAND = 'describe --always';
const BRANCH_COMMAND = 'rev-parse --abbrev-ref HEAD';

function GitInfoPlugin(options = {}) {
  this.worktree = options.worktree;

  this.hashCommand = options.hashCommand || COMMITHASH_COMMAND;

  this.versionCommand = options.versionCommand || VERSION_COMMAND;

  this.branchCommand = options.branchCommand || BRANCH_COMMAND;
}

GitInfoPlugin.prototype.apply = function(compiler) {
  handleCompiler(
    compiler,
    this.worktree,
    this.hashCommand,
    /\[git-hash\]/gi
  );

  handleCompiler(
    compiler,
    this.worktree,
    this.versionCommand,
    /\[git-version\]/gi
  );

  handleCompiler(
    compiler,
    this.worktree,
    this.branchCommand,
    /\[git-branch\]/gi
  );
};

GitInfoPlugin.prototype.hash = function() {
  return runGit(
    this.worktree,
    this.hashCommand
  );
};

GitInfoPlugin.prototype.version = function() {
  return runGit(
    this.worktree,
    this.versionCommand
  );
};

GitInfoPlugin.prototype.branch = function() {
  return runGit(
    this.worktree,
    this.branchCommand
  );
};

module.exports = GitInfoPlugin;
